### PR TITLE
API Deprecate Versioned::canArchive()

### PR DESCRIPTION
**Pull request by GuySartorelli**
For versioned records, there is no concept of "deleting" the record - it's either being archived, unpublished, or in rare cases removed from draft but left published.

Previously the `canDelete()` method has been used to check if a record can be removed from draft specifically for versioned records - which is such a rare edge case it almost doesn't bare thinking about, especially given there's no way to do that purely by interacting with the UI, nor with the web API endpoints available in core or supported modules.

Instead, `canDelete()` should be used to check if the record can be archived. This reduces cognitive load for developers and is a step towards simplifying the overly-complex versioning system. This is also in keeping with other mechanisms, such as `cascade_deletes` which cascades archiving for versioned records.

## Issue
- https://github.com/silverstripe/silverstripe-versioned/issues/447

### CODE REVIEW

File: src/Versioned.php [hunk 1]
```diff
@@ -1498,16 +1498,18 @@ protected function extendCanUnpublish()
...
     {
+        Deprecation::notice('5.3.0', 'Use canDelete() instead.');
         if (!$member) {
             $member = Security::getCurrentUser();
         }
 
         // Standard mechanism for accepting permission changes from extensions
         $owner = $this->owner;
-        $extended = $owner->extendedCan('canArchive', $member);
+        $extended = Deprecation::withNoReplacement(fn() => $owner->extendedCan('canArchive', $member));
```
Comment (ChloeSartorelli) [inline src/Versioned.php:1512]: `extendedCan()` will result in calling `extendCanArchive()` which is also deprecated, so we need to wrap this call too.

File: src/Versioned.php [hunk 2]
```diff
@@ -1530,8 +1532,12 @@ public function canArchive($member = null)
         return true;
     }
 
+    /**
+     * @deprecated 5.3.0 Will be removed without equivalent functionality.
+     */
     protected function extendCanArchive()
     {
+        Deprecation::notice('5.3.0', 'Will be removed without equivalent functionality.');
```
Comment (ChloeSartorelli) [inline src/Versioned.php:1540]: `protected` methods are part of our [definition of public API](https://docs.silverstripe.org/en/5/project_governance/public_api/) so we need to deprecate this even though nobody is likely to be calling it directly and it's effectively a no-op.

File: src/GridFieldArchiveAction.php [hunk 3]
```diff
@@ -176,7 +177,7 @@ public function handleAction(GridField $gridField, $actionName, $arguments, $dat
                 return;
             }
 
-            if (!$item->canArchive()) {
+            if (!Deprecation::withNoReplacement(fn() => $item->canArchive())) {
```
Comment (emteknetnz) [inline src/GridFieldArchiveAction.php]:
```suggestion
            $canArchive = Deprecation::withNoReplacement(fn() => $item->canArchive());
            if (!$canArchive) {
```
Makes this a lot easier to read

Also update all other instances in this PR

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/GridFieldArchiveAction.php (hunk 3 above)
Comment (ChloeSartorelli) [reply src/GridFieldArchiveAction.php]: That's very subjective and I personally don't agree, but I'll make the change to avoid ping pong.

Comment (ChloeSartorelli) [reply src/GridFieldArchiveAction.php]: Done, fingers cross I caught them all

File: src/GridFieldArchiveAction.php [hunk 4]
```diff
@@ -196,7 +198,11 @@ public function handleAction(GridField $gridField, $actionName, $arguments, $dat
     public function getArchiveAction($gridField, $record)
     {
         /* @var DataObject|Versioned $record */
-        if (!$record->hasMethod('canArchive') || !$record->canArchive()) {
+        if (!$record->hasMethod('canArchive')) {
+            return null;
+        }
+        $canArchive = !Deprecation::withNoReplacement(fn() => $record->canArchive());
+        if ($canArchive) {
```
Comment (ChloeSartorelli) [inline src/GridFieldArchiveAction.php]: Had to separate these conditions to comply with https://github.com/silverstripe/silverstripe-versioned/pull/461#discussion_r1716249601

File: src/GridFieldArchiveAction.php [hunk 5]
```diff
@@ -176,7 +177,8 @@ public function handleAction(GridField $gridField, $actionName, $arguments, $dat
                 return;
             }
 
-            if (!$item->canArchive()) {
+            $canArchive = !Deprecation::withNoReplacement(fn() => $item->canArchive());
+            if ($canArchive) {
```
Comment (emteknetnz) [inline src/GridFieldArchiveAction.php]:
```suggestion
            $canArchive = Deprecation::withNoReplacement(fn() => $item->canArchive());
            if (!$canArchive) {
```

File: src/GridFieldArchiveAction.php (hunk 4 above)
Comment (emteknetnz) [inline src/GridFieldArchiveAction.php]:
```suggestion
        $canArchive = Deprecation::withNoReplacement(fn() => $record->canArchive());
        if (!$canArchive) {
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/GridFieldArchiveAction.php (hunk 5 above)
Comment (ChloeSartorelli) [reply src/GridFieldArchiveAction.php]: Oops! Done

File: src/GridFieldArchiveAction.php (hunk 4 above)
Comment (ChloeSartorelli) [reply src/GridFieldArchiveAction.php]: Oops! Done

**Review by emteknetnz — APPROVED** [review]

**Pull request merged**